tests/conftest.py

```diff
@@ -4,8 +4,107 @@
            '<urlopen error [Errno -5] No address associated with hostname>' in str(exception)
+
+
+@pytest.fixture
+def test_dir():
+    return os.path.abspath(os.path.dirname(__file__))
+
+
+def get_url(absolute_path):
+    return urljoin('file:', absolute_path)
+
+
+@pytest.fixture
+def minimal_swagger2dot0_specs_abspath(test_dir):
+    return os.path.join(test_dir, 'data/v2.0/minimal.yaml')
+
+
+@pytest.fixture
+def minimal_swagger2dot0_specs_dict(minimal_swagger2dot0_specs_abspath):
+    return read_file(minimal_swagger2dot0_specs_abspath)
+
+
+@pytest.fixture
+def instagram2dot0_abspath(test_dir):
+    # instragram.json comes from:
+    #
+    # http://editor2.swagger.io/#/
+    #    -> File
+    #       -> Open Example...
+    #          -> instagram.yaml
+    #
+    # and then export it to a json file.
+    return os.path.join(test_dir, 'data/v2.0/instagram.json')
+
+
+@pytest.fixture
+def instagram2dot0_dict(instagram2dot0_abspath):
+    return read_file(instagram2dot0_abspath)
+
+
+@pytest.fixture
+def polymorphic_swagger2dot0_specs_abspath(test_dir):
+    return os.path.join(test_dir, 'data/v2.0/test_polymorphic_specs/swagger.json')
+
+
+@pytest.fixture
+def polymorphic_swagger2dot0_specs_dict(polymorphic_swagger2dot0_specs_abspath):
+    return read_file(polymorphic_swagger2dot0_specs_abspath)
+
+
+@pytest.fixture
+def petstore_swagger2dot0_specs_abspath(test_dir):
+    return os.path.join(test_dir, 'data/v2.0/petstore.json')
+
+
+@pytest.fixture
+def petstore_swagger2dot0_specs_dict(petstore_swagger2dot0_specs_abspath):
+    return read_file(petstore_swagger2dot0_specs_abspath)
+
+
+@pytest.fixture
+def relative_reference_swagger2dot0_specs_abspath(test_dir):
+    return os.path.join(test_dir, 'data/v2.0/relative_ref.json')
+
+
+@pytest.fixture
+def fails_on_invalid_external_ref_swagger2dot0_specs_abspath(test_dir):
+    return os.path.join(test_dir, 'data/v2.0/test_fails_on_invalid_external_ref/petstore.json')
+
+
+@pytest.fixture
+def fails_on_invalid_external_ref_swagger2dot0_specs_dict(fails_on_invalid_external_ref_swagger2dot0_specs_abspath):
+    return read_file(fails_on_invalid_external_ref_swagger2dot0_specs_abspath)
+
+
+@pytest.fixture
+def fails_on_invalid_external_ref_in_list_swagger2dot0_specs_abspath(test_dir):
+    return os.path.join(test_dir, 'data/v2.0/test_fails_on_invalid_external_ref_in_list/petstore.json')
+
+
+@pytest.fixture
+def fails_on_invalid_external_ref_in_list_swagger2dot0_specs_dict(fails_on_invalid_external_ref_in_list_swagger2dot0_specs_abspath):
+    return read_file(fails_on_invalid_external_ref_in_list_swagger2dot0_specs_abspath)
+
+
+@pytest.fixture
+def complicated_refs_swagger2dot0_specs_abspath(test_dir):
+    return os.path.join(test_dir, 'data/v2.0/test_complicated_refs/swagger.json')
+
+
+@pytest.fixture
+def complicated_refs_swagger2dot0_specs_dict(complicated_refs_swagger2dot0_specs_abspath):
+    return read_file(complicated_refs_swagger2dot0_specs_abspath)
```

Comment:
I'm not sure what the advantage is of putting these fixtures in conftest.py. Do you expect to reuse them later, across multiple test files?

Here's how I see it in regards to organizing your code:

1. The closer code is that is "executed together" (i.e. depends on each other, calls each other) the better. Ideally, put it all in one function. This helps readability and reasoning about code.
2. If a function becomes too big or you need to use parts of the logic somewhere else, extract that into a separate function.

Fixtures are a great way to do 2., and conftest.py is a way to share fixtures without having to do imports. Still, the ideal test to me is one where I can immediately see what's mocked and how, what we set as mock results, and what assertions we make. Any time I have to look that up (where's this fixture coming from...) that is a bad thing. It might make sense so we can reuse code and keep files to a reasonable size, but we should always keep in mind what goal we're trying to achieve.

So for fixtures, here's my rule of thumb:

- If that fixture is used by one test (and you don't expect that to change in the near future), do it within the test function. No need to create a separate fixture function.
- If that fixture is used by multiple tests in the same file, put the fixture at the top of the file.
- If you need to share a fixture across multiple test files, conftest.py might be the place to put it.

In summary (and sorry for this long rant), fixture functions and conftest.py are a feature pytest provides to us to organize your code. It doesn't mean at all that every mock and fixture should be in its own function, or that all of these functions should be in conftest.py.

Reply:
Agree ... as we discussed offline the approach was too greedy and lazy :) I moved everything out without double checking if it made sense doing it.
I've refactored the fixtures.